superset/migrations/versions/3325d4caccc8_dashboard_scoped_filters.py

```diff
@@ -0,0 +1,111 @@
+# "License"); you may not use this file except in compliance
+# with the License.  You may obtain a copy of the License at
+#
+#   http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing,
+# software distributed under the License is distributed on an
+# "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+# KIND, either express or implied.  See the License for the
+# specific language governing permissions and limitations
+# under the License.
+"""empty message
+
+Revision ID: 3325d4caccc8
+Revises: e96dbf2cfef0
+Create Date: 2020-02-07 14:13:51.714678
+
+"""
+
+# revision identifiers, used by Alembic.
+import json
+import logging
+
+from alembic import op
+from sqlalchemy import and_, Column, ForeignKey, Integer, String, Table, Text
+from sqlalchemy.ext.declarative import declarative_base
+from sqlalchemy.orm import relationship
+
+from superset import db
+from superset.utils.dashboard_filter_scopes_converter import convert_filter_scopes
+
+revision = "3325d4caccc8"
+down_revision = "e96dbf2cfef0"
+
+Base = declarative_base()
+
+
+class Slice(Base):
+    """Declarative class to do query in upgrade"""
+
+    __tablename__ = "slices"
+    id = Column(Integer, primary_key=True)
+    slice_name = Column(String(250))
+    params = Column(Text)
+    viz_type = Column(String(250))
+
+
+dashboard_slices = Table(
+    "dashboard_slices",
+    Base.metadata,
+    Column("id", Integer, primary_key=True),
+    Column("dashboard_id", Integer, ForeignKey("dashboards.id")),
+    Column("slice_id", Integer, ForeignKey("slices.id")),
+)
+
+
+class Dashboard(Base):
+    __tablename__ = "dashboards"
+    id = Column(Integer, primary_key=True)
+    json_metadata = Column(Text)
+    slices = relationship("Slice", secondary=dashboard_slices, backref="dashboards")
+
+
+def upgrade():
+    bind = op.get_bind()
+    session = db.Session(bind=bind)
+
+    dashboards = session.query(Dashboard).all()
+    for i, dashboard in enumerate(dashboards):
+        print(f"scanning dashboard ({i + 1}/{len(dashboards)}) >>>>")
+        try:
+            json_metadata = json.loads(dashboard.json_metadata or "{}")
+            if "filter_scopes" in json_metadata:
+                continue
+
+            filter_scopes = {}
+            filters = [
+                slice for slice in dashboard.slices if slice.viz_type == "filter_box"
+            ]
+
+            # if dashboard has filter_box
+            if filters:
+                filter_scopes = convert_filter_scopes(json_metadata, filters)
+
+            json_metadata.pop("filter_immune_slices", None)
+            json_metadata.pop("filter_immune_slice_fields", None)
+            if filter_scopes:
+                json_metadata["filter_scopes"] = filter_scopes
+                logging.info(
+                    f"Adding filter_scopes for dashboard {dashboard.id}: {json.dumps(filter_scopes)}"
+                )
+            if json_metadata:
```

Comment:
Don’t we also need to update the JSON metadata of the record if it’s falsey?

Reply:
correct!  I fixed it. 
some old dashboards have no json_metadata. so after this migration i want to keep their json_metadata attribute be `None` instead of `{}`